Update documentation of Random (#19609)
MystenLabs/sui: docs/content/guides/developer/advanced/randomness-onchain.mdx

:::

Having access to random numbers is only one part of designing secure applications, you should also pay careful attention to how you use that randomness.
To securely access randomness:

- Define your function as (private) `entry`.
- Prefer generating randomness using function-local `RandomGenerator`.
- Make sure that the "unhappy path" of your function does not charge more gas than the "happy path".

## Limited resources and `Random` dependent flows

Developers must be aware that some resources that are available to transactions are limited.

> If a function that reads `Random` consumes more resources in an unhappy path flow than in a happy path flow, an attacker can use that difference to revert the transaction in the unhappy flow.

Reviewer: you haven't yet defined "unhappy path" and "happy path" so I might simplify this here to make the warning punchier and more direct. (I think it's fine to keep those terms in the bullet points above, since those are just bullets, but it may be worth explaning in more detail somewhere in the doc what is meant by "happy" and "unhappy")

Just an example for what I might write in place of this line:
```
If you are not careful, an attacker can break or exploit your application by deliberately controlling the point where your function runs out of resources.

Reviewer: to be clear that is what I would replace only line 33 with

Concretely, gas is such a resource.
Consider the following vulnerable code:

```move
// Insecure implementation, do not use!
entry fun insecure_play(r: &Random, payment: Coin<SUI>, ...) {
  ...
  let mut generator = new_generator(r, ctx);
  let win = generator.generate_bool();
  if (win) { // happy flow
    ... cheap computation ...
  } else {
    ... very expensive computation ...
  }
}
```

Observe that the gas costs of a transaction that calls `insecure_play` depends on the value of `win`.
An attacker could call this function with a gas budget that is sufficient for the happy flow but not the unhappy one, resulting in it either winning or reverting the transaction (but never losing the payment).

:::warning

> Sui does not prevent this kind of attack, and developers must be aware of this subtlety when designing their contracts.

Reviewer: optional/personal preference- I would write here "the `random` API" does not automatically prevent this attack, rather than "Sui" doesn't

:::

Other limited resources per transaction that developers should consider are:

- The number of new objects.
- The number of objects that can be used (including dynamic fields).
- Number of events emitted.
- Number of UIDs generated, or deleted, or transferred.

For many use cases this attack is not an issue, like when selecting a raffle winner, or lottery numbers, as the code running is independent of the randomness.
However, in the cases where it can be problematic, you can consider one of the following:

> - Use two steps:

Reviewer: The description here is necessarily a bit abstract - I think you should also link from here directly to the two-step example you wrote in code

Split the logic to two functions that must be called by different transactions.

> The first function, called by transaction `tx1`, fetches a random value and stores it in an object that is unreadable by other commands in `tx1` (e.g., by transferring the object to the caller, or, by storing the tx digest and checking it is different on read).

Reviewer 2: ```suggestion
The first function, called by transaction `tx1`, fetches a random value and stores it in an object that is unreadable by other commands in `tx1` (for example, by transferring the object to the caller, or, by storing the tx digest and checking it is different on read).
```

A second function, called by transaction `tx2`, reads the stored value and completes the operation.
`tx2` might indeed fail, but now the random value is fixed and cannot be modified using repeated calls.
It is important that the inputs to the second function are fixed and cannot be modified after `tx1` (otherwise an attacker can modify them after seeing the randomness committed by `tx1`).

> Also, it is important to gracefully handle the case in which the second step is never completed (e.g., charge a fee in the first step).

Reviewer 2: ```suggestion
Also, it is important to gracefully handle the case in which the second step is never completed (for example, charge a fee in the first step).
```

- Write the function in a way that the happy flow consumes more gas than the unhappy one.
  - Keep in mind that external functions or native ones can change in the future, potentially resulting in different costs compared to the time you conducted your tests.
  - [profile-transaction](../../../references/cli/client.mdx#profile-a-transaction) can be used to profile the costs of a transaction.

## Use (non-public) `entry` functions

While composition is very powerful for smart contracts, it opens the door to attacks on functions that use randomness.

> Consider for example a betting game that uses randomness for rolling a dice:

Reviewer: nit - "rolling a die"

Reviewer 2: ...or "rolling dice"

:::

Having access to random numbers is only one part of designing secure applications, you should also pay careful attention to how you use that randomness.
To securely access randomness:

- Define your function as (private) `entry`.
- Prefer generating randomness using function-local `RandomGenerator`.
- Make sure that the "unhappy path" of your function does not charge more gas than the "happy path".

## Limited resources and `Random` dependent flows

> Developers must be aware that some resources that are available to transactions are limited.

Reviewer 2: ```suggestion
Be aware that some resources that are available to transactions are limited.
```

If you are not careful, an attacker can break or exploit your application by deliberately controlling the point where your function runs out of resources.

Concretely, gas is such a resource.
Consider the following vulnerable code:

```move
// Insecure implementation, do not use!
entry fun insecure_play(r: &Random, payment: Coin<SUI>, ...) {
  ...
  let mut generator = new_generator(r, ctx);
  let win = generator.generate_bool();
  if (win) { // happy flow
    ... cheap computation ...
  } else {
    ... very expensive computation ...
  }
}
```

Observe that the gas costs of a transaction that calls `insecure_play` depends on the value of `win`.
An attacker could call this function with a gas budget that is sufficient for the "happy flow" but not the "unhappy one", resulting in it either winning or reverting the transaction (but never losing the payment).

:::warning

> The `Random` API does not automatically prevent this kind of attack, and developers must be aware of this subtlety when designing their contracts.

Reviewer 2: ```suggestion
The `Random` API does not automatically prevent this kind of attack, and you must be aware of this subtlety when designing your contracts.
```

:::

> Other limited resources per transaction that developers should consider are:

Reviewer 2: ```suggestion
Other limited resources per transaction that you should consider are:
```
